config/v3_0/config_test.go

```diff
@@ -109,3 +110,26 @@ func TestParse(t *testing.T) {
 		assert.Equal(t, test.out.config, config, "#%d: bad config, report: %+v", i, report)
 	}
 }
+
+// TestZeroValueKey walks the types.Config struct hierarchy and verifies
+// that all Key() methods can be called on a zero value without crashing
+// on a nil dereference.  Structure validation is cumulative, so Key
+// functions need to be callable even on an invalid structure.
+func TestZeroValueKey(t *testing.T) {
+	var descend func(typ reflect.Type)
+	descend = func(typ reflect.Type) {
+		switch typ.Kind() {
+		case reflect.Ptr, reflect.Slice:
+			descend(typ.Elem())
+		case reflect.Struct:
+			if key, ok := typ.MethodByName("Key"); ok {
+				instance := reflect.New(typ)
+				key.Func.Call([]reflect.Value{instance.Elem()})
+			}
+			for i := 0; i < typ.NumField(); i++ {
+				descend(typ.Field(i).Type)
+			}
+		}
+	}
+	descend(reflect.TypeOf(types.Config{}))
```

Comment:
phew this is so meta :smile:

One Q: Couldn't we pass the `types.Config{}` as interface parameter to `TestZeroValueKey`, and make it reusable across all versions, instead of implementing it for each?

Reply:
Thanks for raising this.  It turns out we already have a generic structure test, it just hadn't been updated for current specs.  I'll extend that instead.